src/mpi/comm/comm_impl.c

```diff
@@ -726,25 +726,42 @@ static int get_tag_from_stringtag(const char *stringtag)
     return hash % (MPIR_Process.attrs.tag_ub);
 }
 
+static bool is_world_group(MPIR_Group * group_ptr)
+{
+    return (group_ptr->size == MPIR_Process.size && group_ptr->size > 1);
+}
+
+static bool is_self_group(MPIR_Group * group_ptr)
+{
+    return (group_ptr->size == 1);
+}
+
 int MPIR_Comm_create_from_group_impl(MPIR_Group * group_ptr, const char *stringtag,
                                      MPIR_Info * info_ptr, MPIR_Errhandler * errhan_ptr,
                                      MPIR_Comm ** p_newcom_ptr)
 {
     int mpi_errno = MPI_SUCCESS;
+    int use_comm_world = 0;
 
-    /* This implementation assumes an internal MPI_COMM_WORLD already exist.
-     *
-     * Refer to TODOs below inside the branches:
-     * The next implementation will relax but assume the first call to this function will
-     * either from a "world" pset or "self" pset, during former a comm world will be created.
-     *
-     * The final implementation will further relax and allow partial initializations.
-     */
-
-    /* NOTE: tag will be used with MPIR_TAG_COLL_BIT on, ref. MPIR_Get_contextid_sparse_group */
-    int tag = get_tag_from_stringtag(stringtag);
+    /* NOTE: only world or self is supported without first establishing comm_world.  */
 
+    MPL_initlock_lock(&MPIR_init_lock);
     if (MPIR_Process.comm_world) {
+        use_comm_world = 1;
+    } else if (is_world_group(group_ptr)) {
+        mpi_errno = MPIR_init_comm_world();
+        MPIR_ERR_CHECK(mpi_errno);
+        use_comm_world = 1;
```

Comment:
Does this handle process re-ordering?

Reply:
`is_world_group` should return false if the group consists every process but re-ordered.

Reply:
Isn't it just comparing size?
```
static bool is_world_group(MPIR_Group * group_ptr)
{
    return (group_ptr->size == MPIR_Process.size && group_ptr->size > 1);
}
```

Reply:
> `is_world_group` should return false if the group consists every process but re-ordered.
> 
> EDIT: on second thought, we probably can relax that and let `is_world_group` return true as long as every world processes are in the group, reordered or not, since as long as every process is in, we can call `MPIR_init_comm_world` safely. The bootstrapped `MPIR_Process.comm_world` will be used to create the actual communicator, so reordered world group is supported.
> 
> EDIT2: I think the current code behave exactly like that. "Process re-ordering" should work.

OK, I get it now. Thanks.